File: src/libs/base/arch/host/Env.cc
/*
 * Copyright (C) 2015-2018 Nils Asmussen <nils@os.inf.tu-dresden.de>
 * Economic rights: Technische Universitaet Dresden (Germany)
 *
 * Copyright (C) 2019-2022 Nils Asmussen, Barkhausen Institut
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

#include <base/log/Lib.h>
#include <base/Backtrace.h>
#include <base/Env.h>
#include <base/TCU.h>
#include <base/Init.h>
#include <base/Panic.h>

#include <sys/mman.h>
#include <fstream>
#include <unistd.h>
#include <fcntl.h>

#ifndef NDEBUG
volatile int wait_for_debugger = 1;
#endif

namespace m3 {

void *Env::_mem = nullptr;
Env *Env::_inst = nullptr;
INIT_PRIO_ENV Env::Init Env::_init;
char Env::_exec[128];
char Env::_exec_short[128];
const char *Env::_exec_short_ptr = nullptr;

HostEnvBackend::HostEnvBackend() {
}

HostEnvBackend::~HostEnvBackend() {
}

Env::Init::Init() {
#ifndef NDEBUG
    char *val = getenv("M3_WAIT");
    if(val) {
        const char *exec = Env::executable();
        size_t vlen = strlen(val);
        size_t elen = strlen(exec);
        if(elen >= vlen && strcmp(exec + elen - vlen, val) == 0 &&
            (elen == vlen || exec[elen - vlen - 1] == '/')) {
            while(wait_for_debugger)
                usleep(20000);
        }
    }
#endif

    m3::Heap::init();
    Env::init();

    Serial::init(executable(), env()->tile_id);
}

Env::Init::~Init() {
}

const char *Env::tmp_dir() {
    return getenv("M3_HOST_TMP");
}

const char *Env::out_dir() {
    return getenv("M3_OUT");
}

Env::Env(EnvBackend *backend, int logfd)
    : tile_id(set_inst(this)),
      shared(false),
      first_std_ep(0),
      envp(),
      _backend(backend),
      _logfd(logfd),
      _shm_prefix(),
      _sysc_label(),
      _sysc_epid(),
      _sysc_credits(),
      _log_mutex(PTHREAD_MUTEX_INITIALIZER) {
}

void Env::init_executable() {
    int fd = open("/proc/self/cmdline", O_RDONLY);
    if(fd == -1)
        PANIC("open");
    if(read(fd, _exec, sizeof(_exec)) == -1)
        PANIC("read");
    close(fd);
    strncpy(_exec_short, _exec, sizeof(_exec_short));
    _exec_short[sizeof(_exec_short) - 1] = '\0';
    _exec_short_ptr = basename(_exec_short);
}

void *Env::mem() {
    if(_mem == nullptr) {
        _mem = mmap(0, LOCAL_MEM_SIZE, PROT_READ | PROT_WRITE, MAP_ANONYMOUS | MAP_PRIVATE, -1, 0);
        if(_mem == MAP_FAILED)
            PANIC("Unable to map heap");
    }
    return _mem;
}

void Env::print() const {
    char **env = environ;
    while(*env) {
        if(strstr(*env, "M3_") != nullptr || strstr(*env, "LD_") != nullptr) {
            char *dup = strdup(*env);
            char *name = strtok(dup, "=");
            printf("%s = %s\n", name, getenv(name));
            free(dup);
        }
        env++;
    }
}

Env::~Env() {
    delete _backend;
}

}

File: src/libs/rustm3/src/arch/host/env.rs
/*
 * Copyright (C) 2018, Nils Asmussen <nils@os.inf.tu-dresden.de>
 * Economic rights: Technische Universitaet Dresden (Germany)
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

use arch;
use base;
use cap::Selector;
use cell::StaticCell;
use col::{String, Vec};
use com::{SendGate, SliceSource};
use core::intrinsics;
use dtu::{EpId, Label};
use kif::{self, PEDesc, PEType, PEISA};
use libc;
use pes::VPE;
use session::{Pager, ResMng};
use vfs::{FileTable, MountTable};

pub struct EnvData {
    sysc_crd: u64,
    sysc_lbl: Label,
    sysc_ep: EpId,
    _shm_prefix: String,

    vpe: usize,
}

impl EnvData {
    fn base(&self) -> &base::envdata::EnvData {
        base::envdata::get()
    }

    pub fn pe_id(&self) -> u32 {
        self.base().pe_id
    }

    pub fn shared(&self) -> bool {
        self.base().shared != 0
    }

    pub fn pe_desc(&self) -> PEDesc {
        PEDesc::new_from(self.base().pe_desc)
    }

    pub fn argc(&self) -> usize {
        self.base().argc as usize
    }

    pub fn argv(&self) -> *const *const i8 {
        self.base().argv as *const *const i8
    }

    pub fn has_vpe(&self) -> bool {
        self.vpe != 0
    }

    pub fn vpe(&self) -> &'static mut VPE {
        unsafe { intrinsics::transmute(self.vpe) }
    }

    pub fn set_vpe(&mut self, vpe: &VPE) {
        self.vpe = vpe as *const VPE as usize;
    }

    pub fn load_pager(&self) -> Option<Pager> {
        None
    }

    pub fn load_rmng(&self) -> ResMng {
        ResMng::new(SendGate::new_bind(Self::load_word("rmng", 0) as Selector))
    }

    pub fn load_eps(&self) -> u64 {
        Self::load_word("eps", 0)
    }

    pub fn load_nextsel(&self) -> Selector {
        if self.base().first_sel != 0 {
            self.base().first_sel as Selector
        }
        else {
            Self::load_word("nextsel", u64::from(kif::FIRST_FREE_SEL)) as Selector
        }
    }

    pub fn load_rbufs(&self) -> arch::rbufs::RBufSpace {
        match arch::loader::read_env_file("rbufs") {
            Some(rbuf) => {
                let mut ss = SliceSource::new(&rbuf);
                arch::rbufs::RBufSpace::new_with(ss.pop(), ss.pop())
            },
            None => arch::rbufs::RBufSpace::new(),
        }
    }

    pub fn load_mounts(&self) -> MountTable {
        match arch::loader::read_env_file("ms") {
            Some(ms) => MountTable::unserialize(&mut SliceSource::new(&ms)),
            None => MountTable::default(),
        }
    }

    pub fn load_fds(&self) -> FileTable {
        match arch::loader::read_env_file("fds") {
            Some(fds) => FileTable::unserialize(&mut SliceSource::new(&fds)),
            None => FileTable::default(),
        }
    }

    fn load_word(name: &str, default: u64) -> u64 {
        match arch::loader::read_env_file(name) {
            Some(buf) => SliceSource::new(&buf).pop(),
            None => default,
        }
    }

    // --- host specific API ---

    pub fn syscall_params(&self) -> (EpId, Label, u64) {
        (self.sysc_ep, self.sysc_lbl, self.sysc_crd)
    }
}

fn read_line(fd: i32) -> String {
    let mut vec = Vec::new();
    loop {
        let mut buf = [0u8; 1];
        if unsafe { libc::read(fd, buf.as_mut_ptr() as *mut libc::c_void, 1) } == 0 {
            break;
        }
        if buf[0] == b'\n' {
            break;
        }
        vec.push(buf[0]);
    }
    unsafe { String::from_utf8_unchecked(vec) }
}

static ENV_DATA: StaticCell<Option<EnvData>> = StaticCell::new(None);

pub fn get() -> &'static mut EnvData {
    ENV_DATA.get_mut().as_mut().unwrap()
}

pub fn init(argc: i32, argv: *const *const i8) {
    let fd = unsafe {
        let path = format!("/tmp/m3/{}\0", libc::getpid());
        libc::open(path.as_ptr() as *const libc::c_char, libc::O_RDONLY)
    };
    assert!(fd != -1);

    let shm_prefix = read_line(fd);

    let base = base::envdata::EnvData::new(
        read_line(fd).parse::<u32>().unwrap(),
        PEDesc::new(PEType::COMP_IMEM, PEISA::X86, 1024 * 1024),
        argc,
        argv,
        read_line(fd).parse::<Selector>().unwrap(),
        read_line(fd).parse::<Selector>().unwrap(),
    );
    base::envdata::set(base);

    ENV_DATA.set(Some(EnvData {
        sysc_lbl: read_line(fd).parse::<Label>().unwrap(),
        sysc_ep: read_line(fd).parse::<EpId>().unwrap(),
        sysc_crd: read_line(fd).parse::<u64>().unwrap(),
        _shm_prefix: shm_prefix,

        vpe: 0,
    }));

    unsafe {
        libc::close(fd);
    }
}

pub fn reinit() {
    init(get().argc() as i32, get().argv());
}
